### DISCUSSION TITLE
Passing data to a custom loss function

### DISCUSSION THREAD

**Discussion by cacof1** [microsoft/InnerEye-DeepLearning#491] [Q&A] [answered]
Hello all,

Thank you for this nice project!

I am wondering if it's possible to pass patient metadata to the custom get_loss_function defined in the configs/models. I'm thinking specifically of censored data, that I would not want to pass to the model (so non-categorical/numerical features), but that would be useful for my loss function. I would be tempted to include them in the dataset.csv, but I'm not sure how to pass them to my function.

Best,

Charles

**Comment by ant0nsc** [comment, marked as answer]
I'm not sure I understand the question fully. Is the ask to have a loss function that depends on something in the dataset that is specific to each patient? Or something constant across all patients?
It could help if you say a bit more about your data. Is that segmentation? classification?

Update based on discussion: 
I'm afraid the out-of-the-box models will then not work for your use-case. The loss functions that you can create with the `get_loss_function` method will always be called with arguments `(logits, labels)` - not even a subject ID or something alike.
Have you looked into the "bring your own model" framework described [here](https://github.com/microsoft/InnerEye-DeepLearning/blob/main/docs/bring_your_own_model.md)? With that, you would have complete freedom to define custom data loaders, models, and loss functions, while still relying on InnerEye for lifting that all up to Azure. In your custom model, you could re-use, for example, the `image_encoder_with_mlp`.

**Reply by cacof1** [reply]
Thanks for your reply. It is the former of your choices. Here are a few more details. I'm trying to do regression on survival time. For each patient, I have a tag that indicates whether they are lost to follow-up or dead (as a boolean), and I need to integrate that in my loss function. I was trying to write my own CustomRegression loss function through the get_loss_function in my model definition, but I can't figure out how to get that tag through without having it in my model.

**Reply by ant0nsc** [reply]
What's your data like? a plain CSV file? Are there any images? More details would be great to better understand how to tackle that.

**Reply by cacof1** [reply]
The CSV file is the dataset.csv which contains paths to the data and survival time, the data are a series of DICOM-CT/DICOM-RD (one for each patient), that I fed into the image_encoder_with_mlp model.

**Reply by cacof1** [reply]
Thanks for your time! Indeed that seems like the best option, I'll go down that route. Cheers,
